pcmanfm/application.cpp

```diff
@@ -690,6 +699,88 @@ void Application::setWallpaper(QString path, QString modeString) {
     }
 }
 
+/* This method receives a list of file:// URIs from DBus and for each URI opens
+ * a tab showing its content.
+ */
+void Application::ShowFolders(const QStringList uriList, const QString startupId __attribute__((unused))) {
+    if(!uriList.isEmpty()) {
+        launchFiles(QDir::currentPath(), uriList, false, false);
+    }
+}
+
+/* This method receives a list of file:// URIs from DBus and opens windows
+ * or tabs for each folder, highlighting all listed items within each. The
+ * input list is not sorted or grouped so we need to marshal it into groups
+ * by folder, then call our "reveal" method to show each group
+ */
+void Application::ShowItems(const QStringList uriList, const QString startupId __attribute__((unused))) {
```

Comment:
This method has 4 problems: its starting comment has redundant lines, it uses `QFileInfo`, the order of `uriList` isn't kept by it, and its choice of window isn't compatible with that of `launchFiles()`. Please change it to:

```c++
/* This method receives a list of file:// URIs from DBus and opens windows
 * or tabs for each folder, highlighting all listed items within each.
 */
void Application::ShowItems(const QStringList uriList, const QString startupId __attribute__((unused))) {

    QMap<QString,QStringList> groups;
    QStringList keys;
    for(const auto& u : uriList) {
        QString folder = u.section(QLatin1Char('/'), 0, -2);
        if(!folder.isEmpty()) {
            if(!keys.contains(folder)) {
                groups[folder] = QStringList();
                keys << folder; // keep the original order (QMap is sorted by key)
            }
            groups[folder].append(u);
        }
    }

    if(groups.isEmpty()) {
        return;
    }

    PCManFM::MainWindow* window = nullptr;
    if(settings_.singleWindowMode()) {
        window = MainWindow::lastActive();
        if(window == nullptr) {
            QWidgetList windows = topLevelWidgets();
            for(int i = 0; i < windows.size(); ++i) {
                auto win = windows.at(windows.size() - 1 - i);
                if(win->inherits("PCManFM::MainWindow")) {
                    window = static_cast<MainWindow*>(win);
                    break;
                }
            }
        }
    }
    if(window == nullptr) {
        window = new MainWindow();
    }

    for(const auto& k : qAsConst(keys)) {
        window->openFolderAndSelectItems(k, groups[k]);
    }

    window->show();
    window->raise();
    window->activateWindow();
}
```

Reply:
Hey @tsujan I just noticed this
```C++
QString folder = u.section(QLatin1Char('/'), 0, -2);
```
desn't work properly with some URIS, I just put one line to debug cuz I was getting an error when using

`qdbus org.freedesktop.FileManager1 /org/freedesktop/FileManager1 ShowItems "(" "file:///dev" "file:///etc/hosts" ")" "0"`

and it prints this:

![image](https://user-images.githubusercontent.com/32344641/153282274-f3596b92-ccd9-4dfc-8ac9-da74ded2944e.png)

as you can see for the first uri the folder should be "/" (root) but is empty so it gives an error, the other one works properly

Reply:
I'll fix it. Thanks for reporting.

Reply:
The fix is here: https://github.com/lxqt/pcmanfm-qt/pull/1529. We should have worked with `libfm-qt`'s path, instead of manipulating `QString`.

Now, relative paths are also supported, like this from `~/`:
```
qdbus org.freedesktop.FileManager1 /org/freedesktop/FileManager1 ShowItems "(" "Desktop" "./Videos" ")" "1"
```